Add messaging when verified track content MVP is enabled.

cahrens opened the pull request:
## [TNL-4193](https://openedx.atlassian.net/browse/TNL-4393)

This PR adds messaging in the cohorting admin section of the instructor dashboard if the automatic verified track cohorting feature is enabled. See below for screenshots and details.

### Sandbox

- [x] https://cahrens.sandbox.edx.org/courses/course-v1:edx+100+1/instructor#view-cohort_management, can modified verified track settings at https://cahrens.sandbox.edx.org/admin/verified_track_content/verifiedtrackcohortedcourse/ (log in as "admin" with password "edx")

### Testing

- [x] i18n
- [x] Data is properly encoded in HTML templates to avoid XSS
- [x] Unit, integration, acceptance tests as appropriate

Reviewers

If you've been tagged for review, please check your corresponding box once you've given the :+1:.

- [x] Code review: @dianakhuang
- [x] Code review: @efischer19
- [x] UI strings/error msgs review: @catong
- [x] UX review: @chris-mike (optional)
- [ ] Product review: @scottrish

### Post-review

- [x] Squash commits

### Screenshots (all with course that has verified track cohorting enabled, course-v1:edx+100+1):

A) Cohorting not enabled in cohort admin section of instructor dashboard

![image](https://cloud.githubusercontent.com/assets/484484/14942786/0b3b12d4-0f95-11e6-8a0a-e44f5716c3d1.png)

B) Cohorting enabled, no "Verified Learners" cohort exists (note that the ability to disable cohorts is now disabled).

![image](https://cloud.githubusercontent.com/assets/484484/14942789/280bc868-0f95-11e6-9acb-d5f156cdebd8.png)


C) Cohorting enabled, "Verified Learners" cohort exists. In this case, course staff should not be able to rename the Verified Learners cohort name (and also can't disable cohorts).

![image](https://cloud.githubusercontent.com/assets/484484/14942798/6e7c033a-0f95-11e6-9446-471baf2ae4a8.png)

Note that there is an edge case when first creating a cohort with the "verified" name, if it wasn't created before verified track cohorting was enabled (this is NOT the recommended workflow)-- in this edge case, the ability to rename the cohort is not disabled until the cohorting editing template is re-rendered. Given that we do not expect this use case, the cohorting status message DOES update after the rename to state that there is no correctly named cohort, and that the template will re-render on pretty much any other operation afterwards (correcting the issue), I think this is acceptable. I can demonstrate this case if anyone cares. :)

cahrens commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -129,6 +132,19 @@ define(['backbone', 'jquery', 'common/js/spec_helpers/ajax_helpers', 'common/js/
+                }
+                else {
+                    return { enabled: false };
+
```
Remove extra space.

cahrens commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -151,14 +167,28 @@ define(['backbone', 'jquery', 'common/js/spec_helpers/ajax_helpers', 'common/js/
+                flushVerifiedTrackCohortRequests(options.enableVerifiedTrackCohorting);
+            };
+
+            // Flush out all requests to get verified track cohort information.
```
Improve comments.

cahrens commented on lms/templates/instructor/instructor_dashboard_2/cohort-form.underscore:
```diff
@@ -26,7 +26,11 @@
-                           placeholder="<%- placeholder_value %>" required="required" type="text">
+                           placeholder="<%- placeholder_value %>" required="required" type="text"
+                           <% if (cohort.disableEditingName) { %>
+                           readonly
```
Look at getting this to style as readonly.

catong commented on lms/djangoapps/verified_track_content/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+        Verify that cohorting_settings is working for HTTP GET when verified track cohorting is enabled.
+        """
+        config = VerifiedTrackCohortedCourse.objects.create(
+            course_key=unicode(self.course.id), enabled=True, verified_cohort_name="Verified Learners"
```
@cahrens Is this the cohort name suggested by Melanie or Julia? There's no mention of "track" and so there's some possibility of misunderstanding that the learners are in a verified state, rather than that they're on the verified track. Possible to make this "Verified Track Learners" instead?

catong replied:
Oh, is this just in the tests?

cahrens replied:
Just tests. The course authors will determine the name of the "verified track" cohort themselves.

catong replied:
Hm, I got confused about whether courses can use their own name for the verified track cohort. Does the PM specify this in Django Admin and then the name shows up in the UI messages?

cahrens replied:
Yes, that is all correct. This is what the Django admin looks like:

![image](https://cloud.githubusercontent.com/assets/484484/14957585/12cde624-1054-11e6-967e-61a228a60ccb.png)

catong commented on lms/static/js/groups/views/verified_track_settings_notification.js:
```diff
@@ -0,0 +1,79 @@
+                                this.showNotification({
+                                    type: 'confirmation',
+                                    title: StringUtils.interpolate(
+                                        gettext("This course is using automatic cohorting for verified track learners. Cohorts cannot be disabled, and the manual cohort named '{verifiedCohortName}' cannot be renamed. If you wish to change the configuration, contact your PM."), // jshint ignore:line
```
"this course is using" -> "this course uses".
"You cannot disable cohorts, and you cannot rename the manual cohort named ```'{verifiedCohortName}'```. To change the configuration for verified track cohorts, contact your edX partner manager."

cahrens replied:
Done in https://github.com/edx/edx-platform/pull/12278/commits/6168566914b04d83cdc52627ee230ffa29b3da9d.

catong commented on lms/static/js/groups/views/verified_track_settings_notification.js:
```diff
@@ -0,0 +1,79 @@
+                                this.showNotification({
+                                    type: 'error',
+                                    title: StringUtils.interpolate(
+                                        gettext("This course has automatic cohorting enabled for verified track learners, but no manual cohort named '{verifiedCohortName}' exists. You must create '{verifiedCohortName}' for the feature to work."), // jshint ignore:line
```
'This course has automatic cohorting enabled for verified track learners, but the required cohort does not exist. You must create a manually-assigned cohort named ```'{verifiedCohortName}'``` for the feature to work."

cahrens replied:
Done in https://github.com/edx/edx-platform/pull/12278/commits/6168566914b04d83cdc52627ee230ffa29b3da9d.

catong commented on lms/static/js/spec/groups/views/cohorts_spec.js:
```diff
@@ -129,6 +132,18 @@ define(['backbone', 'jquery', 'common/js/spec_helpers/ajax_helpers', 'common/js/
+                if (enabled) {
+                    return {
+                        enabled: true,
+                        verified_cohort_name: "Verified Track"
```
@cahrens Is this the cohort name suggested by Melanie or Julia? There's no mention of "track" and so there's some possibility of misunderstanding that the learners are in a verified state, rather than that they're on the verified track. Possible to make this "Verified Track Learners" instead?

cahrens replied:
This is a test file. The course authors will determine the name of the "verified track" cohort themselves.

efischer19 commented:
Pre-empting @chris-mike a bit here, but the notification colors fail a11y checks for color contrast:

The not-happy case fails with `Element has insufficient color contrast of 2.98 (foreground color: #fd5757, background color: #fff7f7, font size: 12.0pt, font weight: bold)`
<img width="1395" alt="screen shot 2016-05-02 at 10 48 52 am" src="https://cloud.githubusercontent.com/assets/7373924/14957483/92d2a8ba-1053-11e6-9190-23ff3a48b13d.png">

The happy case fails with `Element has insufficient color contrast of 4.34 (foreground color: #008801, background color: #f2f9f2, font size: 12.0pt, font weight: bold)`
<img width="1387" alt="screen shot 2016-05-02 at 10 47 31 am" src="https://cloud.githubusercontent.com/assets/7373924/14957484/92d305f8-1053-11e6-8a7d-8fc8aa981bff.png">

The blue link text for `the Data Downloads Page` in the gray box at the bottom of the page also fails with `Element has insufficient color contrast of 4.31 (foreground color: #0078b0, background color: #f1f1f1, font size: 10.5pt, font weight: normal)`

catong commented on lms/static/js/groups/views/verified_track_settings_notification.js:
```diff
@@ -0,0 +1,79 @@
+                        else {
+                            this.showNotification({
+                                type: 'error',
+                                title: gettext('This course has automatic cohorting enabled for verified track learners, but cohorts are disabled. You must enable cohorts for the feature to work.') // jshint ignore:line
```
This string looks good.

efischer19 commented on lms/static/js/groups/views/verified_track_settings_notification.js:
```diff
@@ -0,0 +1,79 @@
+                                    cohort.get('name') === verifiedCohortName) {
+                                    verifiedCohortExists = true;
+                                    cohort.disableEditingName = true;
+                                }
```
Nit: The indentation of this if-block's contents lining up with the last part of the condition threw me for a second on first read, could the body of the block get bumped inwards one more tab?

efischer19 commented:
One minor nit and the a11y failures noted above, otherwise 👍 from me on the code and functionality. I'm okay with the edge case you described (assuming product is as well).

catong commented:
👍 from me for the UI strings with my edits in comments.

cahrens commented:
FYI @chris-mike and others who care, the instructor dashboard messaging is odd in that the warning and error colors being used (for "warning" type notifications and "error" type notifications) are both variants of red. I would have expected "warning" to be an orangish color.

Here is an example of both being shown at the same time (with a modified red to pass accessibility audit). The top message is of type "error", while the bottom is of type "warning".

![image](https://cloud.githubusercontent.com/assets/484484/14959750/d2a4f06e-105e-11e6-9c48-46d0cd046cae.png)

cahrens commented:
@efischer19 Thanks for finding the contrast issues. I fixed the two related to notifications, but didn't touch the one Data Downloads because that grey background color is used throughout the page (for the cohort settings tab background), and there are many contrast issues pre-existing on this page.

I have updated the sandbox and will add a few comments on my cleanup commit.

cahrens commented on lms/static/sass/base/_variables.scss:
```diff
@@ -201,7 +201,7 @@ $shadow-d2: rgba($black, 0.6) !default;
 $error-color: rgb(253, 87, 87) !default;
 $warning-color: rgb(181,42,103) !default;
-$confirm-color: rgb(0, 136, 1) !default;
+$confirm-color: rgb(0, 132, 1) !default;
```
$confirm-color is only used by the instructor dashboard CSS, and only a minor tweak was needed to get sufficient contrast.

cahrens commented on lms/static/sass/course/instructor/_instructor_2.scss:
```diff
@@ -84,11 +84,11 @@
   .msg-error {
-    border-top: 2px solid $error-color;
-    background: tint($error-color,95%);
+    border-top: 2px solid $red1;
```
@error-color is used in many places throughout the product, so I didn't feel comfortable changing it. There is a "newer" block of colors that had $red1, so I switched to that. I was also able to make the background slightly darker (90% tint) without running afoul of contrast violations-- to me, that looks better.

Note though the issue I already mentioned-- $warning-color is being used for "warning" messages, but it's a dark red (magenta) color, which is odd.

cahrens commented:
@chris-mike I'm not sure if you care to review this PR-- it's using existing notification patterns in the cohorting management section of the instructor dashboard. I did have to tweak some colors though for contrast ("confirmation" green and "error" red), and as I noted, it's odd that the warning messages (I have not introduced any new warning messages... this is a pre-existing issue) are using a red color.

dianakhuang commented on lms/djangoapps/verified_track_content/tests/test_views.py:
```diff
@@ -0,0 +1,73 @@
+    """
+    Tests the `cohort_discussion_topics` view.
+    """
+    SAMPLE_COURSE = 'edX/Test_Course/Run'
```
This doesn't seem to be used anywhere, and it would make more sense to use `self.course.id` anyway (which you already seem to be doing).

cahrens replied:
Right-- left over from an earlier version of the test. I will delete!

dianakhuang commented:
Other than that one minor comment, I think it looks good. 👍

cahrens commented:
jenkins run python

cahrens commented:
jenkins run lettuce

cahrens commented:
jenkins run bokchoy

robrap commented:
@cahrens The following is being added as a requirement to our TNL template:
- [ ] ./scripts/safe-commit-linter.sh shows 0 violations
You may want to preemptively run this on this PR and see what happens.

chris-mike commented:
Thanks @efischer19 for doing the web aim check. I might have designed these error messages differently, to be more consistent with visual design in the pattern library (http://ux.edx.org/). We acknowledge four levels of "alert": error (red), success (green), warning (yellow) and information (blue-grey:#6FA0BA).  

To your point @cahrens , the magenta color should be closer to a warning, which is #FDBC56.  Can we change magenta to yellow in this field, similar to what is designed in pattern lib? Or is this a different issue?

cahrens commented:
@chris-mike I didn't touch anything related to warnings in this PR, so it is out-of-scope. I'd make the change anyway if it only impacted the instructor dashboard, but $warning-color (the coloring being referenced) is used in a number of different places-- courseware, system-feedback, fields, courseware_preview, etc.

I have no idea why $warning-color was ever made its current color, but it looks like it has been this way since Brian introduced it on 9/17/2015 as part of comprehensive theming.

cahrens commented:
@chris-mike That being said, I can try changing the instructor dashboard usage to $warning-base, which looks like it is a yellowish color taken from the pattern library. Let me see if the contrast is sufficient....

chris-mike commented:
Thanks, @cahrens . I would use the white background, and the yellow bar. Another helpful affordance would be the icon preceeding the message (font awesome:  fa-warning (alias) [&#xf071;] ).

cahrens commented:
@chris-mike It doesn't have sufficient contrast, even with a white background. I don't think that color was ever intended as a text color.

![image](https://cloud.githubusercontent.com/assets/484484/14967347/d82c1e7e-1084-11e6-9c4c-30a533d3d085.png)

I don't wish to get into changing the notification area (adding an icon) for this PR.

cahrens commented:
@robrap The linter found a few issues. I fixed one of them-- https://github.com/edx/edx-platform/pull/12278/commits/0f8b24523c2e493d3ec18e78fa8ab7c71dd2a5cd. The others are a bit complicated, and I don't feel equipped to fix them right now by myself. Perhaps we can sit together and do it tomorrow.

chris-mike commented:
Look at the alerts pattern in ux.edx.org. The yellow is only used in the bar color and in the icon. The text is the standard charcoal-ish black.

cahrens commented:
@chris-mike Yes, I see that. But I feel that changing the notification UI is out-of-scope for this story.

cahrens commented:
jenkins run js

robrap commented:
@cahrens: I'd be happy to help on linter issues.

cahrens commented:
@chris-mike Here is the tech debt story for the notification message work. I'll add some more specific details to it later today about what needs to be done. https://openedx.atlassian.net/browse/TNL-4486

@scottrish gave the OK to merge this without his review.
